### PR TITLE
feat(address): add payment method billing details

### PR DESCRIPTION
**Pull request by Narayanbhat166**
## Type of Change

- [x] New feature

## Description

This PR adds the ability to pass `billing` details in `payment_method_data`. This feature is required in order to pass the whole of billing data in `payment_method_data` instead of passing only required fields, as is the current implementation. 

### Additional Changes

- [x] This PR modifies the API contract
- [x] This PR modifies the database schema

## Motivation and Context

This PR is a part of refactoring billing details in payment method data. 

## Impact
Payment method data api contract

## How did you test it?

- Create a payment with `payment_method_data.billing`, below curl can be used.
```bash
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'api-key: <API_KEY>' \
--data-raw '{
    "amount": 6969,
    "currency": "USD",
    "confirm": true,
    "capture_method": "automatic",
    "setup_future_usage": "off_session",
    "description": "Its my fourth payment request",
    "authentication_type": "no_three_ds",
    "payment_method": "card",
    "payment_method_type": "credit",
    "payment_method_data": {
        "card": {
            "card_number": "4242424242424242",
            "card_exp_month": "10",
            "card_exp_year": "25",
            "card_holder_name": "joseph Doe",
            "card_cvc": "123"
        },
        "billing": {
            "address": {
                "line1": "1467",
                "line2": "Harrison Street",
                "line3": "Harrisoff Street",
                "city": "San Fransico",
                "state": "California",
                "zip": "94122",
                "country": "US",
                "first_name": "Narayan",
                "last_name": "Doe"
            },
            "email": "example@juspay.in"
        }
    },
    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        }
    },
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrisoff Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "Narayan",
            "last_name": "Doe"
        },
        "email": "example@juspay.in"
    }
}'
```
- Ch
Screenshot 2024-02-26 at 2 32 06 PM
eck if it is received in the response. ( under `payment_method_data.billing` )

- Check if it is stored in database, get the `payment_method_billing_address_id` from `payment_attempt` and fetch the address.
image

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code

### CODE REVIEW

File: crates/router/src/core/payments/operations/payment_update.rs [hunk 1]
```diff
@@ -193,6 +198,21 @@ impl<F: Send + Clone, Ctx: PaymentMethodRetrieve>
         )
         .await?;
 
+        let payment_method_billing = helpers::create_or_update_address_for_payment_by_request(
+            db,
+            request.billing.as_ref(),
```
Comment (sai-harsha-vardhan) [inline crates/router/src/core/payments/operations/payment_update.rs:203]: Shouldn't we take this from payment_method_data.billing?

Comment (Narayanbhat166) [reply crates/router/src/core/payments/operations/payment_update.rs:203]: Good catch, this must be taken from request.payment_method_data. will change it. Will also add test cases covering this

**Review by sai-harsha-vardhan — APPROVED** [review]

**Review by hrithikesh026 — APPROVED** [review]
LGTM

File: crates/router/src/core/payments/operations/payment_confirm.rs [hunk 2]
```diff
@@ -479,6 +543,9 @@ impl<F: Send + Clone, Ctx: PaymentMethodRetrieve>
             address: PaymentAddress {
                 shipping: shipping_address.as_ref().map(|a| a.into()),
                 billing: billing_address.as_ref().map(|a| a.into()),
+                payment_method_billing: payment_method_billing
+                    .as_ref()
+                    .map(|address| address.into()),
```
Comment (SanchithHegde) [inline crates/router/src/core/payments/operations/payment_confirm.rs:548]:
```suggestion
                    .map(Into::into),
```

**Review by SanchithHegde — APPROVED** [review]

**Pull request merged**